h2/src/main/org/h2/mvstore/db/LobStorageMap.java

```diff
@@ -250,75 +270,102 @@ public boolean isReadOnly() {
     public ValueLob copyLob(ValueLob old_, int tableId, long length) {
-        init();
-        ValueLobDatabase old = (ValueLobDatabase) old_;
-        int type = old.getValueType();
-        long oldLobId = old.getLobId();
-        long oldLength = old.getType().getPrecision();
-        if (oldLength != length) {
-            throw DbException.getInternalError("Length is different");
-        }
-        Object[] value = lobMap.get(oldLobId);
-        value = value.clone();
-        byte[] streamStoreId = (byte[]) value[0];
-        long lobId = generateLobId();
-        value[1] = tableId;
-        lobMap.put(lobId, value);
-        Object[] key = { streamStoreId, lobId };
-        refMap.put(key, Boolean.TRUE);
-        ValueLob lob = ValueLobDatabase.create(
-                type, database, tableId, lobId, length);
-        if (TRACE) {
-            trace("copy " + old.getTableId() + "/" + old.getLobId() +
-                    " > " + tableId + "/" + lobId);
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            ValueLobDatabase old = (ValueLobDatabase) old_;
+            int type = old.getValueType();
+            long oldLobId = old.getLobId();
+            long oldLength = old.getType().getPrecision();
+            if (oldLength != length) {
+                throw DbException.getInternalError("Length is different");
+            }
+            Object[] value = lobMap.get(oldLobId);
+            value = value.clone();
+            byte[] streamStoreId = (byte[]) value[0];
+            long lobId = generateLobId();
+            value[1] = tableId;
+            lobMap.put(lobId, value);
+            Object[] key = {streamStoreId, lobId};
+            refMap.put(key, Boolean.TRUE);
+            ValueLob lob = ValueLobDatabase.create(
+                    type, database, tableId, lobId, length);
+            if (TRACE) {
+                trace("copy " + old.getTableId() + "/" + old.getLobId() +
+                        " > " + tableId + "/" + lobId);
+            }
+            return lob;
+        } finally {
+            deregisterVersionUsage(txCounter);
         }
-        return lob;
     }
 
     @Override
     public InputStream getInputStream(long lobId, long byteCount)
             throws IOException {
-        init();
-        Object[] value = lobMap.get(lobId);
-        if (value == null) {
-            throw DbException.get(ErrorCode.LOB_CLOSED_ON_TIMEOUT_1, "" + lobId);
+        MVStore.TxCounter txCounter = initialize();
+        try {
+            Object[] value = lobMap.get(lobId);
+            if (value == null) {
+                throw DbException.get(ErrorCode.LOB_CLOSED_ON_TIMEOUT_1, "" + lobId);
+            }
+            byte[] streamStoreId = (byte[]) value[0];
+            InputStream inputStream = streamStore.get(streamStoreId);
+            return new FilterInputStream(inputStream) {
+                @Override
+                public int read(byte[] b, int off, int len) throws IOException {
```

Comment:
How about `read(byte[])`, `read()`, `skip()`? They can be used through JDBC, even more, JDBC layer can use `skip()` by itself.

Reply:
`read(byte[])` calls `read(byte[], int, int)`, so it should be fine. `read()` does some work by itself, but I think initialization of `oneByteBuffer` doesn't need to be wrapped into version, so it should be fine too. `skip()` also doesn't read anything.

Other methods from `InputStream` don't have custom implementations anyway.

Reply:
FilterInputStream.read(byte[]) calls read(byte[],int,int) and skip does not do actual map access, therefore does not need a protection, but read() is my oversight indeed. I forgot that FilterInputStream is a wrapper, not a subclass.

Reply:
I missed that `StreamStore.read()` calls unwrapped `StreamStore.read(byte[], int, int)` too.